great_expectations/cli/datasource.py

```diff
@@ -25,167 +25,176 @@ def add_datasource(context):
-                                        show_default=True)
-                password = click.prompt("What is the password for the sqlalchemy connection?", default="",
-                                        show_default=False, hide_input=True)
-                database = click.prompt("What is the database name for the sqlalchemy connection?", default="postgres",
-                                        show_default=True)
-
-                # Since we don't want to save the database credentials in the config file that will be
-                # committed in the repo, we will use our Variable Substitution feature to store the credentials
-                # in the credentials file (that will not be committed, since it is in the uncommitted directory)
-                # with the datasource's name as the variable name.
-                # The value of the datasource's "credentials" key in the config file (great_expectations.yml) will
-                # be ${datasource name}.
-                # GE will replace the ${datasource name} with the value from the credentials file in runtime.
-                credentials = {
-                    "drivername": drivername,
-                    "host": host,
-                    "port": port,
-                    "username": username,
-                    "password": password,
-                    "database": database
-                }
-            else:
-                sqlalchemy_url = click.prompt(
-"""What is the url/connection string for the sqlalchemy connection?
-(reference: https://docs.sqlalchemy.org/en/latest/core/engines.html#database-urls)
-""",
-                                    show_default=False)
-                credentials = {
-                    "url": sqlalchemy_url
-                }
-
-            context.save_config_variable(data_source_name, credentials)
-
-            try:
-                context.add_datasource(data_source_name,
-                                       module_name="great_expectations.datasource",
-                                       class_name="SqlAlchemyDatasource",
-                                       data_asset_type={"class_name": "SqlAlchemyDataset"},
-                                       credentials="${" + data_source_name + "}")
-                break
-            except (DatasourceInitializationError, ModuleNotFoundError) as de:
-                cli_message(
-"""
-Cannot connect to the database. Please check your environment and the configuration you provided.
-
-<red>Actual error: {0:s}</red>>
-""".format(str(de)))
-                if not click.confirm(
-"""
-Enter the credentials again?
-""".format(str(de)),
-                    default=True):
-                    cli_message(
-"""
-Exiting datasource configuration.
-You can add a datasource later by editing the great_expectations.yml file.
-""")
-                    return None
-
-
+        data_source_name = _add_sqlalchemy_datasource(context)
```

Comment:
Factored this giant method down into a few smaller data source specific chunks. The diff looks massive, but it's really actually quite small.